tool/fragment/src/main/java/org/openscience/cdk/fragment/FunctionalGroupsFinder.java

```diff
@@ -0,0 +1,1339 @@
+     *
+     * @param aMolecule the molecule to identify functional groups in
+     * @param aShouldInputBeCloned use 'false' to reuse the input container's bonds and atoms in the extraction of the functional
+     *                             groups; this may speed up the extraction and lower the memory consumption for processing large
+     *                             amounts of data but corrupts the original input container; use 'true' to work with a clone and
+     *                             leave the input container intact
+     * @throws CloneNotSupportedException if cloning is not possible
+     * @return a list with all functional groups found in the molecule
+     */
+    public List<IAtomContainer> find(IAtomContainer aMolecule, boolean aShouldInputBeCloned) throws CloneNotSupportedException {
+        return this.find(aMolecule, aShouldInputBeCloned, false);
+    }
+    //
+    /**
+     * Find all functional groups in a molecule.
+     *
+     * @param aMolecule the molecule to identify functional groups in
+     * @param aShouldInputBeCloned use 'false' to reuse the input container's bonds and atoms in the extraction of the functional
+     *                             groups; this may speed up the extraction and lower the memory consumption for processing large
+     *                             amounts of data but corrupts the original input container; use 'true' to work with a clone and
+     *                             leave the input container intact
+     * @param anAreInputRestrictionsApplied if true, the input must consist of one connected structure and may not
+     *                                      contain charged atoms, metals or metalloids; an IllegalArgumentException will
+     *                                      be thrown otherwise; see convenience methods in this class for detecting
+     *                                      illegal input structures for this case
+     * @throws CloneNotSupportedException if cloning is not possible
+     * @throws IllegalArgumentException if input restrictions are applied and the given molecule does not fulfill them
+     * @return a list with all functional groups found in the molecule
+     */
+    public List<IAtomContainer> find(IAtomContainer aMolecule, boolean aShouldInputBeCloned, boolean anAreInputRestrictionsApplied)
+            throws CloneNotSupportedException, IllegalArgumentException {
+        this.clearCache();
+        IAtomContainer tmpMolecule;
+        if (aShouldInputBeCloned) {
+            tmpMolecule = aMolecule.clone();
+        } else {
+            tmpMolecule = aMolecule;
+        }
+        for (IAtom tmpAtom : tmpMolecule.atoms()) {
+            if(Objects.isNull(tmpAtom.getImplicitHydrogenCount())) {
+                tmpAtom.setImplicitHydrogenCount(0);
+            }
+        }
+        this.bondMapCache = EdgeToBondMap.withSpaceFor(tmpMolecule);
+        this.adjListCache = GraphUtil.toAdjList(tmpMolecule, this.bondMapCache);
```

Comment:
You probably don't need this anymore since AtomContainer2 provides these optimisations... but not wrong do this.